hashes/src/lib.rs

```diff
@@ -186,6 +186,9 @@ pub type HkdfSha512 = Hkdf<sha512::Hash>;
 
 /// A hashing engine which bytes can be serialized into.
 pub trait HashEngine: Clone {
+    /// The `Hash` type returned when finalizing this engine.
+    type Hash: Hash;
```

Comment:
If you're going to mess with the traits here it should be `GeneralHash` (but I think it shouldn't be required anyway).

Reply:
`GeneralHash` is folded into `Hash` and deleted here.

As for why do we need the bound -- I tried removing it and it broke the HmacEngine which wants to be able to finalize its inner engine (obtaining a `HashEngine::Hash` type) and then call `.as_ref` to convert it to bytes so that it can iterate over it and do some xor magic.

I think we should leave this be and revisit this in a followup, since it's a nontrivial change.

Reply:
IIRC I've originally suggested `AsRef<[u8]>` bound which should work for the vast majority of use cases (but I wouldn't mind also having `Borrow`, `Copy` and equality traits).

Reply:
Can we merge as is and follow up? I did try various things yesterday and couldn't get anything nicer to work.

Reply:
I could be wrong because every experiment requires loads of boring changes but I definitely tried `AsRef<[u8]>` trait bound.

Reply:
Oh, I think I see your problems. One is that you're using derives inappropriately, the other is that you're, confusingly, using `Hash` in `Hmac` type parameter but `HashEngine` in `Hkdf` type parameter. I'm not sure if it's literally bad though.

Reply:
I'm not really sure if we should merge it in this state. The code before was not horribly broken, so clean history has value and the changes I have in mind touch a significant chunk of code, I think.

Reply:
I think we should merge it. This gets rid of the `GeneralHash` trait and greatly reduces the amount of stuff we need to think about (we have two interacting traits, `Hash` and `HashEngine`, instead of three, and it seems like we're moving toward also getting rid of `Hash`.)

Even if this means that we have a series of big diffs, conceptually it'll be easier to follow removing the big `GeneralHash` trait in one go, then reducing or removing `Hash` in another go.

Reply:
OK, I think I'll be able to review properly tomorrow.

Reply:
Crap, so many things slipping by